tokio-util/src/codec/framed_impl.rs

```diff
@@ -120,42 +120,101 @@ where
+            // returned eof, `decode_eof` will attemp to decode the remainder as closing frames.
+            // The buffer might become readable again if the underlying AsyncRead is a FIFO or file.
+            // However, we still want to make sure that upon encountering an eof
+            // we actually finish emmiting all it's associated `decode_eof` frames.
+            // Furthermore, we don't want to emit any `decode_eof` frames on retried
+            // reads after an eof unless we've actually read more data.
             if state.is_readable {
+                // pausing or framing
                 if state.eof {
+                    // pausing
                     let frame = pinned.codec.decode_eof(&mut state.buffer)?;
+                    if frame.is_none() {
+                        state.is_readable = false; // prepare pausing -> paused
+                    }
+                    // implicit pausing -> pausing or pausing -> paused
                     return Poll::Ready(frame.map(Ok));
                 }
 
+                // framing
                 trace!("attempting to decode a frame");
 
                 if let Some(frame) = pinned.codec.decode(&mut state.buffer)? {
                     trace!("frame decoded from buffer");
+                    // implicit framing -> framing
                     return Poll::Ready(Some(Ok(frame)));
                 }
 
+                // framing -> reading
                 state.is_readable = false;
             }
-
-            assert!(!state.eof);
-
-            // Otherwise, try to read more data and try again. Make sure we've
-            // got room for at least one byte to read to ensure that we don't
-            // get a spurious 0 that looks like EOF
-            state.buffer.reserve(1);
+            // reading or stopped
+            // If we can't build a frame yet, try to read more data and try again.
+            // Make sure we've got room for at least one byte to read to ensure
+            // that we don't get a spurious 0 that looks like eof.
+            // Only allocate if we actually need it to prevent slowly leaking bytes,
+            // in less fortunate code that repeatedly polls an eof file.
+            if state.buffer.capacity() == state.buffer.len() {
+                state.buffer.reserve(1);
+            }
             let bytect = match poll_read_buf(pinned.inner.as_mut(), cx, &mut state.buffer)? {
                 Poll::Ready(ct) => ct,
+                // implicit reading -> reading or implicit stopped -> stopped
                 Poll::Pending => return Poll::Pending,
             };
             if bytect == 0 {
+                if state.eof {
+                    // We're already at an eof, and since we've reached this path
+                    // we're also not readable. This implies that we've already finished
+                    // our `decode_eof` handling, so we can simply return `None`.
+                    // implicit stopped -> stopped
```

Comment:
Stopped? Your state diagram has no `stopped` state.

Reply:
Whops, sorry, I should not work so late, that should be paused, I'm gonna fix it in the morning, to not do the same mistake twice 😆

Reply:
Fixed!